### PR TITLE
feat(bigquery): move the bigquery backend back into the main ibis repo

### PR DESCRIPTION
**Pull request by cpcloud**
This PR moves bigquery back into the main ibis repo.

~Still working through the failing tests, though many are fixed.~ Tests are passing.

TODOs:

- [x] get all tests passing locally
- [x] see if we can automatically handle the autonaming we're doing that bigquery doesn't accept
- [x] setup ci similar to ibis-bigquery if possible, though maybe we just run these tests on push events only similar to snowflake

Possible follow ups:

- [x] delete <4 legacy code
- [x] delete <4 legacy tests
- [x] move SQL tests to write-then-compare so that they are easy to modify

### CODE REVIEW

File: .github/workflows/ibis-backends-cloud.yml
Comment (mik-laj) [inline .github/workflows/ibis-backends-cloud.yml:72]:
```suggestion
        run: echo "SNOWFLAKE_URL=${SNOWFLAKE_URL}" >> "$GITHUB_ENV"
        env:
            SNOWFLAKE_URL: ${{ secrets.SNOWFLAKE_URL }}
```
We should use with 'env' option to avoid script injection attacks or injection problems. See: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections

To set the environment variable, you should also pass the environment variable name. https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable

File: ibis/tests/sql/conftest.py
Comment (mik-laj) [inline ibis/tests/sql/conftest.py]: What do you think about using the `pytest-snapshot `package to write these tests? 
Thanks to this, we will be able to update the tests with one command.
```
pytest --snapshot-update.
```
We also won't have to have a separate file for each assertion, but we will have more assertions in one snapshot file.

Comment (cpcloud) [reply ibis/tests/sql/conftest.py]: I put up #4836 to do this for our existing sqlgolden checks. I'll merge that PR when it's green.

**Comment by jreback** [comment]
@cpcloud sure 

is this a general change in policy though? or a specific one off for BQ?

eg what about a lot of the other google variants or mssql for example

File: ibis/backends/base/sql/registry/window.py
Comment (kszucs) [inline ibis/backends/base/sql/registry/window.py:59]: During the refactors my general experience was that `getattr` this way made it harder to locate type problems by swallowing attribute errors. 
Based on the function's name `_replace_interval_with_scalar` I'd expect to always have an interval dtype here, are other inputs allowed?

File: ibis/backends/base/sql/registry/window.py
Comment (kszucs) [inline ibis/backends/base/sql/registry/window.py:94]: Do we expect `ops.SortKey` instances here? If yes could we use `x.expr` instead?

I know that the window._order_by member is not properly validated yet.

File: ibis/backends/base/sql/registry/window.py
Comment (kszucs) [inline ibis/backends/base/sql/registry/window.py:227]: You should be able to use `c.output_shape.is_columnar()` here.

Comment (kszucs) [reply ibis/backends/base/sql/registry/window.py:227]: Perhaps a similar `c.output_dtype.is_temporal()` convenience method would be useful too.

File: ibis/backends/bigquery/backports.py
Comment (kszucs) [inline ibis/backends/bigquery/backports.py:1]: Since the backend will be tied with the corresponding ibis core revision now, do we still need this backports file?

File: ibis/backends/base/sql/registry/window.py
Comment (cpcloud) [reply ibis/backends/base/sql/registry/window.py:94]: `ops.SortKey`s are allowed, to support `DESC`

File: .github/workflows/ibis-backends-cloud.yml
Comment (cpcloud) [reply .github/workflows/ibis-backends-cloud.yml]: Fixed, thanks!

File: ibis/tests/sql/conftest.py
Comment (cpcloud) [reply ibis/tests/sql/conftest.py]: #4836 is merged!

File: ibis/backends/base/sql/registry/window.py
Comment (cpcloud) [reply ibis/backends/base/sql/registry/window.py:59]: Other inputs are allowed, e.g. in `2 * my_interval` the recursion will eventually hit the `2` which will have an `int8` dtype.

File: ibis/backends/bigquery/__init__.py
Comment (kszucs) [reply ibis/backends/bigquery/__init__.py:235]: Great to know, thanks Gil!

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:235]: These warnings are correct, because we don't run bigquery tests until a merge to `master`.

File: ibis/backends/base/sql/registry/window.py
Comment (cpcloud) [reply ibis/backends/base/sql/registry/window.py]: This was done in #4841.

File: ibis/backends/bigquery/client.py
Comment (kszucs) [inline ibis/backends/bigquery/client.py:42]: Please create a follow-up issue to factor the datatype conversions out to `ibis.backends.bigquery.datatypes`

File: ibis/backends/bigquery/compiler.py
Comment (kszucs) [inline ibis/backends/bigquery/compiler.py:116]: `nodes = lin.traverse(find_bigquery_udf, expr, filter=ops.Node)` should be sufficient

File: ibis/backends/bigquery/client.py
Comment (cpcloud) [reply ibis/backends/bigquery/client.py:42]: Would you mind if I did that in a follow-up?

File: ibis/backends/bigquery/compiler.py
Comment (cpcloud) [reply ibis/backends/bigquery/compiler.py:116]: Yeah, this was leftover from an attempted update to support 4.0.0 in the external backend. I'll fix this up.

**Review by kszucs — APPROVED** [review]
Just nits, other than those it's good to go from my side

File: ibis/backends/bigquery/client.py
Comment (kszucs) [reply ibis/backends/bigquery/client.py:42]: Not at all, just have an issue to track this

**Comment by cpcloud** [comment]
@tswast Would you mind having a look here?

There are a lot of files, but most of the new files are generated SQL tests to make it easier to update expected code when tests break but don't change semantics, so you can likely ignore those.

I was able to either preserve each test as it was in `ibis-bigquery` or generalize it and move to a backend test, so that should give us some confidence about functionality going forward.

File: ibis/backends/bigquery/compiler.py
Comment (cpcloud) [reply ibis/backends/bigquery/compiler.py]: Fixed!

**Comment by cpcloud** [comment]
Ok, this is no longer blocked on https://github.com/googleapis/python-bigquery/pull/1413, I set `python = "<3.11"` for the `google-cloud-bigquery` dependency, which we can remove once that PRs commit is in a release.

File: ibis/backends/bigquery/client.py
Comment (cpcloud) [reply ibis/backends/bigquery/client.py:42]: Done: https://github.com/ibis-project/ibis/issues/4881

File: ibis/backends/bigquery/tests/unit/udf/test_core.py
Comment (tswast) [inline ibis/backends/bigquery/tests/unit/udf/test_core.py:25]: Are these tests ones that we'd want to parametrize with `snapshot` as well?

**Review by tswast — COMMENTED** [review]
BQ changes LGTM. I like the "snapshot" structure in the tests.

File: ibis/backends/bigquery/tests/unit/udf/test_core.py
Comment (cpcloud) [reply ibis/backends/bigquery/tests/unit/udf/test_core.py:25]: Yeah, I think so!

Comment (cpcloud) [reply ibis/backends/bigquery/tests/unit/udf/test_core.py]: Done.

File: ibis/backends/bigquery/__init__.py
Comment (cpcloud) [inline ibis/backends/bigquery/__init__.py:353]: @tswast Can you take a look at this block of code and say whether this is expected behavior?

The use case is reading from in `bigquery-public-data.hacker_news.comments`, but having `ibis-gbq` be the billing project.

Without this workaround, the storage API creates a read session in the data project (`bigquery-public-data`), which causes queries to fail when using the pyarrow functionality.

Comment (tswast) [reply ibis/backends/bigquery/__init__.py:353]: I wouldn't expect this to be necessary. If the query succeeded, then I assume the billing project is being set correctly in the client constructor https://github.com/ibis-project/ibis/blob/01bd4026e463984ab9fd17ea6d57ee6e42d50b8b/ibis/backends/bigquery/__init__.py#L169 and in the query method https://github.com/ibis-project/ibis/blob/01bd4026e463984ab9fd17ea6d57ee6e42d50b8b/ibis/backends/bigquery/__init__.py#L236

I've filed https://github.com/googleapis/python-bigquery/issues/1422 to investigate this further, but I think it's fine to keep this workaround if there really is a bug.

Comment (tswast) [reply ibis/backends/bigquery/__init__.py:353]: Update: I think there really is a bug. The project from "client" is used instead of the project from the QueryJob.

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:353]: Ah, okay. Thanks Tim. I'll keep this comment unresolved so the link is easier to find.

**Comment by cpcloud** [comment]
Ok, I'm going to merge this in and fix any issues with the CI. Thanks all for the help reviewing, great to see this back in the main repo!

**Review by mik-laj — APPROVED** [review]

**Pull request merged**
